### PR TITLE
SBR calculation pipeline

### PR DESCRIPTION
**Pull request by charlotte-avery**
Wrote SBR calculation pipeline based on ESC's calculations: https://docs.google.com/spreadsheets/d/1LqUzYlYcvkqyx9c1_ZokSeoSzK7HsX9g/edit#gid=1184908035  
Ian should use `main.py` script to output SBR, SBR values and Flex Archetype for a given set of user inputs.

### CODE REVIEW

**Review by enyciaa — APPROVED** [review]
This looks great! 

Could the README be updated to explain how to use the library & what it does? 
e.g. inputs, available functions, outputs

File: src/smart_building_rating_calculator/flex_archetype.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+class FlexArchetypes(StrEnum):
+    NO_FLEXER = "No Flexer"
+    LOW_TECH_FLEXER = "Low-tech Flexer"
+    GOOD_FLEXER = "Good Flexer"
+    UNTAPPED_FLEXER = "Untapped Flexer"
+    STRONG_FLEXER = "Strong Flexer"
+    GOLD_FLEXER = "Gold Standard Flexer"
+
+
+def calc_flex_archetype(user_inputs: UserInputs, sbr_val: float) -> FlexArchetypes:
```
Comment (anguschadney) [inline src/smart_building_rating_calculator/flex_archetype.py]: Something to consider here would be to instantiate the `UserInputs` dataclass for each flex archetype and then check for equality (which dataclasses handle [out of the box](https://khuyentran1401.github.io/Efficient_Python_tricks_and_tools_for_data_scientists/Chapter2/dataclasses.html#compare-between-two-data-classes))...

E.g.
```
no_flexer = UserInputs(True, True, False, True ...)
if user_inputs == no_flexer:
  return FlexArchetypes.NO_FLEXER.value
```

You could also then use a `match ... case` statement to simplify further

File: src/smart_building_rating_calculator/main.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    HotWaterSource,
+)
+from src.smart_building_rating_calculator.scoring import (
+    calc_electrification_score,
+    calc_ics_score,
+)
+from src.smart_building_rating_calculator.flex_archetype import calc_flex_archetype
+
+
+def calc_sbr(sbr_val: float) -> str:
```
Comment (anguschadney) [inline src/smart_building_rating_calculator/main.py]: adds pandas which is a bit of a burden for just this job but you can use `pd.cut` to do this in a one-liner...

https://pandas.pydata.org/docs/reference/api/pandas.cut.html

File: src/smart_building_rating_calculator/scoring.py [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+
+
+def calc_v2g_score(user_inputs: UserInputs) -> float:
+    if user_inputs.smart_v2g_enabled:
+        return 3
+    else:
+        return 0
+
+
+def calc_home_battery_score(user_inputs: UserInputs) -> float:
```
Comment (anguschadney) [inline src/smart_building_rating_calculator/scoring.py]: Again you might want static `HomeBattery` classes for all the different instances which you can map to an output value, it might make the logic a bit more clear / easy to maintain

File: src/smart_building_rating_calculator/scoring.py [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+        and user_inputs.battery_size == BatterySize.STANDARD
+        and (not user_inputs.smart_v2g_enabled)
+        and (not user_inputs.smart_ev_charger)
+    ):
+        return 3.5
+    else:
+        return 0
+
+
+def calc_solar_pv_score(user_inputs: UserInputs) -> float:
```
Comment (anguschadney) [inline src/smart_building_rating_calculator/scoring.py]: Same with this one!

File: src/smart_building_rating_calculator/scoring.py [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+    ] or user_inputs.hot_water_source in [
+        HotWaterSource.HEAT_BATTERY_OR_ELEC_HOT_WATER_TANK,
+        HotWaterSource.ELEC_SHOWER_BOILER_OR_OTHER,
+    ]:
+        return -1
+    else:
+        return 0
+
+
+def calc_home_heating_score(user_inputs: UserInputs) -> float:
```
Comment (anguschadney) [inline src/smart_building_rating_calculator/scoring.py]: And so on...

File: tests/test_sbr_calculator.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    EVChargerPower,
+    BatterySize,
+    SolarInverterSize,
+    HeatingSource,
+    HotWaterSource,
+)
+from src.smart_building_rating_calculator.flex_archetype import FlexArchetypes
+from src.smart_building_rating_calculator.main import sbr_score
+
+
```
Comment (anguschadney) [inline tests/test_sbr_calculator.py:35]: I would test the individual scoring functions as well as the archetypes, otherwise you might miss edge cases (e.g. the archetypes do not cover all the possible combinations)

P.s. copilot will generate all of them for you if you write a comment like:
`# Test all combinations of score_...`

**Review by anguschadney — COMMENTED** [review]
great work but some things to think about!

File: src/smart_building_rating_calculator/flex_archetype.py (hunk 1 above)
Comment (shengy90) [reply src/smart_building_rating_calculator/flex_archetype.py]: Alternative implementation to Gus's suggestion - abstract each 'if conditions' into a separate function:

- This long if-conditions are very long and hard to read and going to be hard to unit test
- Would be easier to abstract each 'checks' into a function and unit test each function individually
- You also won't need long if-elif-else since you're returning at the end of each condition. You could just:

```python
if check_gold_flexer(user_inputs):
     return GOLD_FLEXER.value

if check_strong_flexer(user_inputs):
    return STRONG_FLEXER.value
```

etc. That makes this function a lot smaller, testable, and readable.

File: src/smart_building_rating_calculator/main.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    battery_size: BatterySize,
+    solar_pv: bool,
+    pv_inverter_size: SolarInverterSize,
+    electric_heating: bool,
+    heating_source: HeatingSource,
+    hot_water_source: HotWaterSource,
+    secondary_heating: bool,
+    secondary_hot_water: bool,
+    integrated_control_sys: bool,
+):
```
Comment (shengy90) [inline src/smart_building_rating_calculator/main.py]: Return typing:

```python

from typing import Tuple

def test_fun(x: int) -> Tuple[int, int, int]:
    return 1,2,3
```

Something like that - can't remember exact syntax but that's how you can type multiple returns.

File: src/smart_building_rating_calculator/scoring.py [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+        v2g_score,
+        home_battery_score,
+        solar_pv_score,
+        elec_heating_score,
+        home_heating_score,
+        alternative_heating_score,
+        hot_water_heating_score,
+        alternative_hot_water_score,
+    ]
+    print(elec_scores)
```
Comment (shengy90) [inline src/smart_building_rating_calculator/scoring.py]: Avoid print statements in function, or add more context to print statements (e.g. print(f"Electrification score: {score}").

File: tests/test_sbr_calculator.py (hunk 6 above)
Comment (shengy90) [reply tests/test_sbr_calculator.py:35]: Agreed - related to my above feedback of either abstracting out the scoring logic and test it one by one etc.

File: tests/test_sbr_calculator.py [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+def test_high_sbr_gold_flexer():
```
Comment (shengy90) [inline tests/test_sbr_calculator.py]: All these tests are testing the same function. I'd wrap them up in 1 test class:

class Test_sbr_score:
    def test_gold_flexer():
    def test_strong_flexer():

etc

Comment (shengy90) [reply tests/test_sbr_calculator.py]: You should also run tests for the intermediate functions in between - and not rely on just a single test on the final function, i.e. 'unit' test!

File: src/smart_building_rating_calculator/main.py (hunk 7 above)
Comment (shengy90) [reply src/smart_building_rating_calculator/main.py]: Also have a think about the user experience with this function. For someone that has pip-installed this library, what's their experience like calling this function? At the moment, they have to run _a lot_ of import statements just to be able to call this function.

It'd be much easier for this function to take `UserInput` class the input. Have a separate function that takes in 'user inputs' to instantiate the `UserInput` class. Also have a think of how user will be creating the inputs.

For web clients, these would likely be in a JSON object/ string/ int values. A wrapper function to convert this string inputs into UserInput object which is then parsed into this function would be more convenient for end users.

You can then also write unit tests to test that inputs are parsed properly.

Implementation could be very similar to Gus's feedback on "instantiate the UserInputs dataclass for each flex archetype and then check for equality".

Comment (shengy90) [reply src/smart_building_rating_calculator/main.py]: I would also avoid naming this script `main.py`.

`main.py` or `app.py` is usually reserved for an application where `main.py`/`app.py` are the entry points to said app.

I would call this script "calculate_sbr_score.py`.

Think about the common python libraries that you use - I don't think you'll have ever run `from package.main import fn`!

File: README.md [hunk 10]
```diff
@@ -3,4 +3,4 @@ The calculation to generate a smart building rating
 
 ## Setup
 - Run `make setup` which will run all the required commands
-- Run `pipenv run python smart-building-rating-calculator/main.py`
\ No newline at end of file
+- Run `pipenv run python smart-building-rating-calculator/main.py`
```
Comment (shengy90) [inline README.md]: Add some end-to-end examples of how to run the function, especially if this repo is to be made public!

**Comment by shengy90** [comment]
Agreed with Gus's comments! Also I think we should break the functions down a bit more and write proper tests. Right now we're only testing 'end-to-end' which isn't quite 'unit test'!

File: setup.py [hunk 11]
```diff
@@ -1,9 +1,9 @@
-from setuptools import setup, find_packages
+from setuptools import find_packages, setup
 
 setup(
```
Comment (shengy90) [inline setup.py:3]: Here you're using `find_packages` in setup.py - this will basically find all importable modules in your package and be installed by end-users which isn't necessarily what we want. In this case it will actually include your `test` folder as you have an `__init__.py` in there.

It's better to list explicitly what you want to be installed as part of the package (i.e. only codes on `setup.py`). You can use `package_dir={'': 'src'}` to do that, or exclude test in find_packages. I'd prefer to explicitly only include things in `src` to be installed.

You don't want the `tests` folder (or other dev stuffs) to be installed by your end users. You generally only want the actual `package` installed only (i.e. stuff in `src` folder).

See this stackoverflow:
https://stackoverflow.com/questions/14454744/does-it-make-sense-to-install-my-python-unit-tests-in-site-packages

File: src/smart_building_rating_calculator/main.py (hunk 2 above)
Comment (charlotte-avery) [reply src/smart_building_rating_calculator/main.py]: Done

File: src/smart_building_rating_calculator/flex_archetype.py (hunk 1 above)
Comment (charlotte-avery) [reply src/smart_building_rating_calculator/flex_archetype.py]: In reference to @anguschadney's suggestion: Some archetypes depend on `heating_source` being one out of multiple options or `hot_water_source` having a specific value, e.g., to be a "Good Flexer" the household has to have either an EV, battery, HP or electric storage heater. This would mean I would have 4 x options for UserInputs for the user to be a "Good Flexer", i.e.,  good_flexer_1 = UserInputs(...), good_flexer_2 = UserInputs(...), good_flexer_3=, etc. For other flex types, there are even more ways of being that type of flexer. So this makes the suggested logic become quite complex as I would have to write out all the possible options for the user inputs for each archetype. 

To simplify & clarify the code I have instead taken @shengy90's suggestion here. For the functions in `scoring.py`, however, I have taken your suggested approach and created static classes for all the different instances which map to output values.

File: src/smart_building_rating_calculator/scoring.py (hunk 3 above)
Comment (charlotte-avery) [reply src/smart_building_rating_calculator/scoring.py]: Done.

File: src/smart_building_rating_calculator/scoring.py (hunk 4 above)
Comment (charlotte-avery) [reply src/smart_building_rating_calculator/scoring.py]: Done.

File: src/smart_building_rating_calculator/scoring.py (hunk 5 above)
Comment (charlotte-avery) [reply src/smart_building_rating_calculator/scoring.py]: Done.

File: src/smart_building_rating_calculator/scoring.py (hunk 8 above)
Comment (charlotte-avery) [reply src/smart_building_rating_calculator/scoring.py]: Done.

File: tests/test_sbr_calculator.py (hunk 6 above)
Comment (charlotte-avery) [reply tests/test_sbr_calculator.py:35]: Added tests to implement this - let me know if tests can be better written (the test all combinations functions is quite complex)

File: tests/test_sbr_calculator.py (hunk 9 above)
Comment (charlotte-avery) [reply tests/test_sbr_calculator.py]: Added more extensive testing functions & units tests.

File: src/smart_building_rating_calculator/main.py (hunk 7 above)
Comment (charlotte-avery) [reply src/smart_building_rating_calculator/main.py]: Renamed the script to `calculate_sbr_score.py` as recommended and reduced the number of imports required to run script.

File: README.md (hunk 10 above)
Comment (charlotte-avery) [reply README.md]: Done.

File: setup.py (hunk 11 above)
Comment (charlotte-avery) [reply setup.py:3]: Good spot! I have now directly specified the src directory for `find_packages`.

File: src/smart_building_rating_calculator/calculate_sbr_score.py [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+        Tuple[float, str, str]: SBR value, SBR rating, Flex Archetype
+    """
+    sbr_val = calc_sbr_score(user_inputs)
+    sbr = calc_sbr(sbr_val)
+    flex_archetype = calc_flex_archetype(user_inputs, sbr_val)
+
+    return sbr_val, sbr, flex_archetype
+
+
+def sbr_score(
```
Comment (shengy90) [inline src/smart_building_rating_calculator/calculate_sbr_score.py:52]: Missing types

File: tests/test_sbr_calculator.py [hunk 13]
```diff
@@ -0,0 +1,353 @@
...
+    calc_electrification_score,
+    calc_ics_score,
+    calc_smart_meter_score,
+)
+
+
+class TestScoreCalculators:
+    """Test the scoring calculations using the results in the SBR excel sheet created by ESC"""
+
+    def test_smartest_home(self):
```
Comment (shengy90) [inline tests/test_sbr_calculator.py]: The proper way to do this is to parametrize the test by:

1. Creating these UserInputs as pytest fixtures
2. Call these fixtures and parametrize the test and compare it against expected results

https://engineeringfordatascience.com/posts/pytest_fixtures_with_parameterize/

File: tests/test_sbr_calculator.py [hunk 14]
```diff
@@ -0,0 +1,353 @@
...
+            solar_pv=True,
+            pv_inverter_size=SolarInverterSize.LT_4KW,
+            electric_heating=True,
+            heating_source=HeatingSource.HEAT_PUMP,
+            hot_water_source=HotWaterSource.HEAT_BATTERY_OR_ELEC_HOT_WATER_TANK,
+            secondary_heating=True,
+            secondary_hot_water=True,
+            integrated_control_sys=True,
+        )
+        elec_scores = calc_electrification_score(user_inputs)
```
Comment (shengy90) [inline tests/test_sbr_calculator.py]: You should be unit testing these functions. This test right now is testing way too many things at once. If there's a bug in one of this function, this test would fail without giving developers any clue as to where to even begin troubleshooting.

File: tests/test_sbr_calculator.py [hunk 15]
```diff
@@ -0,0 +1,353 @@
...
+        smart_meter_score = calc_smart_meter_score(user_inputs)
+        ics_score = calc_ics_score(user_inputs)
+        elec_scores.insert(0, smart_meter_score)
+        elec_scores.insert(len(elec_scores), ics_score)
+
+        expected_scores = [1, 3, 3, 1, 3, -1, 4, 2, 1, 1, 1.5]
```
Comment (shengy90) [inline tests/test_sbr_calculator.py]: The purpose of unit tests is not just to test correct input = correct output, but also to add syntax to explain to people what the tests are actually checking.

Writing expected scores like these are not giving developers any syntax on what these numbers mean or why we should be expecting this results.

File: tests/test_sbr_calculator.py [hunk 16]
```diff
@@ -0,0 +1,353 @@
...
+        for score, expected_val in zip(elec_scores, expected_scores):
+            assert score == expected_val
+
+        sbr_val, sbr, flex_archetype = get_sbr_scores(user_inputs)
+        assert sbr_val == 1
+        assert sbr == "G"
+        assert flex_archetype == FlexArchetype.LOW_TECH_FLEXER
+
+
+def test_all_combinations():
```
Comment (shengy90) [inline tests/test_sbr_calculator.py]: I'm not sure I like this test:

1. It's not clear what this test is trying to test for
2. If you're writing that many levels of nested for/if statements, that should be a bit of an alarm bell already!
3. Unit test is meant to test 1 small thing at a time to provide clues to where bugs are. This test is testing a lot of things at once that if broken, doesn't actually provide clues to developers where to debug, i.e. not serving its purpose.


If for some reason you really need to do that many for loops, a more pythonic way is to use itertools.product to get the cartesian product of lists:
https://note.nkmk.me/en/python-itertools-product/

File: src/smart_building_rating_calculator/calculate_sbr_score.py (hunk 12 above)
Comment (charlotte-avery) [reply src/smart_building_rating_calculator/calculate_sbr_score.py:52]: Done

File: tests/test_sbr_calculator.py (hunk 13 above)
Comment (charlotte-avery) [reply tests/test_sbr_calculator.py]: Nice, thanks, I have now implemented fixtures as arguments.

File: tests/test_sbr_calculator.py (hunk 16 above)
Comment (charlotte-avery) [reply tests/test_sbr_calculator.py]: This test is to check that all possible inputs are converted to valid outputs. This test actually flagged some bugs in the code where some combinations of inputs gave weird outputs. so was a helpful test! I will remove for loops for cartesian products.

File: tests/test_sbr_calculator.py (hunk 15 above)
Comment (charlotte-avery) [reply tests/test_sbr_calculator.py]: Let me know if better now please. Thanks!

File: tests/test_sbr_calculator.py (hunk 14 above)
Comment (charlotte-avery) [reply tests/test_sbr_calculator.py]: Let me know if better now please. Thanks!

File: tests/test_sbr_calculator.py (hunk 15 above)
Comment (charlotte-avery) [reply tests/test_sbr_calculator.py]: Note that I am taking the expected scores from ESC's spreadsheet - I don't know what the numbers mean or why we should be expecting this results, other than I expect these values based on the calculations ESC have provided!

File: README.md [hunk 17]
```diff
@@ -1,6 +1,72 @@
...
+3) Flex Archetype (see `src/smart_building_rating_calculator/flexer_enums.py`).
+
+Inputs must have datatypes as defined in `src/smart_building_rating_calculator/inputs.py`
+- Most inputs are `bool` type (True/False)
+- Others are StrEnum type e.g., `charger_power` must have a value of `EVChargerPower("3 kW")`, `EVChargerPower("7 kW")`, `EVChargerPower("22 kW")`, or `EVChargerPower("None")`
+- Upon calling `sbr_score`, correct input datatypes are automatically checked for. An error is raised if input datatypes are incorrect.
+
+Example of how to call `sbr_score` in python:
+
+```
```
Comment (shengy90) [inline README.md]: You can add `python` after ``` to add Python syntax highlighting in Python:
https://www.markdownguide.org/extended-syntax/#syntax-highlighting

File: tests/test_sbr_calculator.py [hunk 18]
```diff
@@ -0,0 +1,524 @@
...
+        heating_source=HeatingSource.HEAT_PUMP,
+        hot_water_source=HotWaterSource.HEAT_BATTERY_OR_ELEC_HOT_WATER_TANK,
+        secondary_heating=True,
+        secondary_hot_water=True,
+        integrated_control_sys=False,
+    )
+    return user_inputs
+
+
+@pytest.mark.parametrize(
```
Comment (shengy90) [inline tests/test_sbr_calculator.py:205]: Non-blocking: You could also test for negative cases too and check that errors are expected:
https://docs.pytest.org/en/8.1.x/how-to/skipping.html#skip-xfail-with-parametrize

File: tests/test_sbr_calculator.py [hunk 19]
```diff
@@ -0,0 +1,524 @@
...
+
+        if (
+            (not inputs[1] and inputs[2] != EVChargerPower.NONE)
+            or (not inputs[4] and inputs[5] != BatterySize.NONE)
+            or (not inputs[6] and inputs[7] != SolarInverterSize.NONE)
+            or (inputs[1] and inputs[2] == EVChargerPower.NONE)
+            or (inputs[4] and inputs[5] == BatterySize.NONE)
+            or (inputs[6] and inputs[7] == SolarInverterSize.NONE)
+        ):
+            with pytest.raises(AssertionError):
```
Comment (shengy90) [inline tests/test_sbr_calculator.py]: It's unclear exactly what this assertion test is testing for by looking at this code. Looking at the function `sbr_score`, that function is not running any assertion tests.

So it's not clear why I'm expecting certain inputs to be failing here. If assertion tests are failing even further upstream of `sbr_score`, then why are we not testing at source but nesting assertion tests so deeply?

I had to trace all the functions called by `sbr_score` to actually find out assertion tests are carried out in `prep_user_input`. These assertion tests should be written as a test to test `prep_user_input` directly instead of having these tests here.

The whole point of unit tests is so that when test fails, we can quickly identify where the source of bug is. Right now it's not apparent.

File: tests/test_sbr_calculator.py [hunk 20]
```diff
@@ -0,0 +1,524 @@
...
+                sbr_normalised, sbr, flex_archetype = sbr_score(*inputs)
+
+        else:
```
Comment (shengy90) [inline tests/test_sbr_calculator.py]: A better way to write this test is to wrap this up as a pytest Test class:

```python
class TestAllCombination:
    @pytest.mark.xfail(raises=AssertionError)
    def test_raises_exceptions_for_bad_inputs(self):
        """run logic that will fail test"""
    def test_assert_expected_types(self):
        assert isinstance(x, float), "x is not float"
        assert isinstance(x, str), "x is not str"

    def test_assert_value_in_list(self):
         assert x in y, "x must be in y"
         etc.
```

This breaks down the test into individual tests, each tests doing 1 thing, so that when things break you can easily find trace back to the source where bugs occur and fixes.
IDE

Doing that also makes unit testing work with many IDEs test plugins to quickly navigate failing tests and find out why tests are failing.

The purpose of unit tests is not just for one to check logic, it's also to guard against regression, i.e. if a new developer comes in and make changes and accidentally break codes, it helps them to quickly find out where things are breaking and make fixes.

**Review by shengy90 — APPROVED** [review]
Approving to unblock but left non-blocking comments on tests.

File: tests/test_sbr_calculator.py (hunk 19 above)
Comment (charlotte-avery) [reply tests/test_sbr_calculator.py]: Makes sense, thanks Sheng. I now test on `prep_user_input` directly

File: tests/test_sbr_calculator.py (hunk 20 above)
Comment (charlotte-avery) [reply tests/test_sbr_calculator.py]: Hopefully better now!
